crates/containerd-shim-wamr/src/instance.rs

```diff
@@ -0,0 +1,97 @@
+use anyhow::{Context, Result};
+use containerd_shim_wasm::container::{Engine, Entrypoint, Instance, RuntimeContext, Stdio};
+use wamr_rust_sdk::function::Function;
+use wamr_rust_sdk::instance::Instance as WamrInstnace;
+use wamr_rust_sdk::module::Module;
+use wamr_rust_sdk::runtime::Runtime;
+use wamr_rust_sdk::wasi_context::WasiCtxBuilder;
+
+pub type WamrInstance = Instance<WamrEngine>;
+
+pub struct WamrEngine {
+    runtime: Runtime,
+}
+
+unsafe impl Send for WamrEngine {}
+unsafe impl Sync for WamrEngine {}
+
+// TODO: wasmr_rust_sdk::runtime::Runtime should implement Clone
+
+impl Default for WamrEngine {
+    fn default() -> Self {
+        let runtime = Runtime::new().unwrap();
+        Self { runtime }
+    }
+}
+
+impl Clone for WamrEngine {
+    fn clone(&self) -> Self {
+        let runtime = Runtime::new().unwrap();
+        Self { runtime }
+    }
+}
+
+impl Engine for WamrEngine {
+    fn name() -> &'static str {
+        "wamr"
+    }
+
+    fn run_wasi(&self, ctx: &impl RuntimeContext, stdio: Stdio) -> Result<i32> {
+        let args = ctx.args();
+        let envs = ctx.envs();
+        let Entrypoint {
+            source,
+            func,
+            arg0: _,
+            name,
+        } = ctx.entrypoint();
+
+        let wasm_bytes = source
+            .as_bytes()
+            .context("Failed to get bytes from source")?;
+
+        log::info!("Create a WAMR module");
+
+        // TODO: error handling isn't ideal
+
+        let mod_name = name.unwrap_or_else(|| "main".to_string());
+
+        let mut module = Module::from_buf(&self.runtime, &wasm_bytes, &mod_name).map_err(|e| {
+            anyhow::Error::msg(format!("Failed to create module from bytes: {:?}", e))
+        })?;
+
+        log::info!("Create a WASI context");
+
+        let wasi_ctx = WasiCtxBuilder::new()
+            .set_pre_open_path(vec!["/"], vec![])
+            .set_env_vars(envs.iter().map(String::as_str).collect())
+            .set_arguments(args.iter().map(String::as_str).collect())
+            .build();
+
+        module.set_wasi_context(wasi_ctx);
+
+        // TODO: no way to register a named module with bytes?
+
+        log::info!("Create a WAMR instance");
+
+        let instance = WamrInstnace::new(&self.runtime, &module, 1024 * 64)
+            .map_err(|e| anyhow::Error::msg(format!("Failed to create instance: {:?}", e)))?;
+
+        log::info!("redirect stdio");
+        stdio.redirect()?;
+
+        log::info!("Running {func:?}");
+        let function = Function::find_export_func(&instance, &func)
+            .map_err(|e| anyhow::Error::msg(format!("Failed to find function: {:?}", e)))?;
+        let status = function
+            .call(&instance, &vec![])
+            .map(|_| 0)
+            .or_else(|err| {
+                log::error!("Error: {:?}", err);
+                Err(err)
+            })
```

Comment:
we should parse the RuntimeError to appropriate exit code. Any thoughts on this one, @lum1n0us?

Reply:
Regarding that, I just noticed that WAMR only uses error strings to convey information if something goes wrong during the execution of a Wasm function. So, if it's okay, can we continue to display the error string in the log and use a simple -1 and 0 to indicate failure and success, respectively?

Reply:
I was looking for exit code that represents https://github.com/containerd/runwasi/blob/main/crates/containerd-shim-wasmer/src/tests.rs#L79-L88

Reply:
For example, how can I get exit code 42 from running this wasm module?

```wat
(module
    ;; Import the required proc_exit WASI function which terminates the program with an exit code.
    ;; The function signature for proc_exit is:
    ;; (exit_code: i32) -> !
    (import "wasi_snapshot_preview1" "proc_exit" (func $proc_exit (param i32)))
    (memory 1)
    (export "memory" (memory 0))
    (func $main (export "_start")
        (call $proc_exit (i32.const 42))
        unreachable
    )
)

```

Reply:
We have [a fix](https://github.com/bytecodealliance/wamr-rust-sdk/pull/79) for this situation. Please have a look.

Reply:
Approved

Reply:
Merged